DigitalLearningSolutions.Data/Helpers/LearningHubSsoSecurityHelper.cs

```diff
@@ -0,0 +1,68 @@
+﻿namespace DigitalLearningSolutions.Data.Helpers
+{
+    using System;
+    using System.Security.Cryptography;
+    using System.Text;
+    using DigitalLearningSolutions.Data.Services;
+
+    public interface ILearningHubSsoSecurityHelper
+    {
+        string GenerateHash(string state, string secretKey);
+
+        bool VerifyHash(string state, string secretKey, string hash);
+    }
+
+    public class LearningHubSsoSecurityHelper : ILearningHubSsoSecurityHelper
+    {
+        private static readonly DateTime UnixEpoch = new DateTime(1970, 1, 1, 0, 0, 0, DateTimeKind.Utc);
+
+        private readonly IClockService clockService;
+
+        public LearningHubSsoSecurityHelper(IClockService clockService)
+        {
+            this.clockService = clockService;
+        }
+
+        public string GenerateHash(string state, string secretKey)
+        {
+            var secondsSinceEpoch = GetSecondsSinceEpoch();
+            var encoder = new UTF8Encoding();
+            var salt = encoder.GetBytes(secretKey);
+            var timedState = encoder.GetBytes(state + secondsSinceEpoch);
+            return GetHash(timedState, salt);

```

Comment:
It does not say in the documentation but I imagine the output needs to be url-encoded.
Is that something for this helper class to do or the service that uses this helper to do?

Reply:
I think that should be in the service calling the helper, on grounds that the helper should only perform its specific task of creating or verifying hashes. If those hashes need to be sent somewhere else or further processed in some way then that's the job of the rest of the system.

Reply:
Sounds reasonable to me. Url encoding will probably be another helper if we even need one.